### PR TITLE
[Core/Metrics] Accurately record used memory.

### LINKED ISSUE #24624: [Core] `ray_object_store_available_memory` metric goes negative when fallback allocation is used
We calculate the available object store memory for this metric via `config_.object_store_memory - used_memory_` [here](https://github.com/ray-project/ray/blob/6e17c4a2b71eeeb556a638ae09d379af862b79f9/src/ray/object_manager/object_manager.cc#L739), yet `used_memory_` can exceed the configured object store capacity via fallback allocation to disk since we always increase this [on the seal path](https://github.com/ray-project/ray/blob/6e17c4a2b71eeeb556a638ae09d379af862b79f9/src/ray/object_manager/object_manager.cc#L195). This results in the `ray_object_store_available_memory` metric going negative, which can be confusing to users. We should change this metric calculation to `config_.object_store_memory - (used_memory_ - plasma::plasma_store_runner->GetFallbackAllocated())`.

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?

used_memory_ includes the fallback allocation, so we should subtract it here to calculate the exact available memory.

## Related issue number

Closes https://github.com/ray-project/ray/issues/24624

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/object_manager/object_manager.cc [hunk 1]
```diff
@@ -736,7 +736,11 @@ std::string ObjectManager::DebugString() const {
 void ObjectManager::RecordMetrics() {
   pull_manager_->RecordMetrics();
   push_manager_->RecordMetrics();
-  stats::ObjectStoreAvailableMemory().Record(config_.object_store_memory - used_memory_);
+  // used_memory_ includes the fallback allocation, so we should subtract it here
+  // to calculate the exact available memory.
+  stats::ObjectStoreAvailableMemory().Record(
+      config_.object_store_memory - used_memory_ -
```
Comment (scv119) [inline src/ray/object_manager/object_manager.cc:742]: should it be `config_.object_store_memory - used_memory_ + plasma::plasma_store_runner->GetFallbackAllocated()` ?

Comment (clarkzinzow) [reply src/ray/object_manager/object_manager.cc:742]: Yep it should be `config_.object_store_memory - used_memory_ - plasma::plasma_store_runner->GetFallbackAllocated()` or `config_.object_store_memory - (used_memory_ + plasma::plasma_store_runner->GetFallbackAllocated())`

Comment (clarkzinzow) [reply src/ray/object_manager/object_manager.cc:742]: Yep it should be `config_.object_store_memory - used_memory_ - plasma::plasma_store_runner->GetFallbackAllocated()` or `config_.object_store_memory - (used_memory_ + plasma::plasma_store_runner->GetFallbackAllocated())`

**Comment by rkooo567** [comment]
Added tests. I also realize we need to handle `used_memory` properly too (if total - used is not avail, it is pretty weird), so I handled it and added a proper test

**Review by clarkzinzow — APPROVED** [review]
Nice, great test!

**Review by scv119 — APPROVED** [review]

**Pull request merged**